Ensemble algorithm and code update for corrected data release

lmcarboneau opened the pull request:
**Please don't merge this yet! It needs a code review, especially for the files outside of the ensemble!**

There are changes to the code that I'm concerned might conflict across methods, including:

1. updating the `taskmanager` to accept a user specified data source (to force FFI preference)

2. updating `run_tesscorr` to accept a user specified data source; updating to also accept a user specified `output_folder,` and to pass a user specified debug flag to the `BaseCorrector`

3. updating the `BaseCorrector` to accept the new parameters above, and implementing internal changes to use them properly (init: `self.output_folder = output_folder`;  `save_lightcurve()` and plots now save to `self.output_folder` when specified, etc)

4. updating the BaseCorrector to use `lc_corr` instead of `lc` (original object) :: there was some bug that occasionally didn't make a new object when it wasn't implicit, but it also increases the readability of the code in this section

5. passing `Tmag` along to the ensemble corrector :: I haven't checked if quality flags are also being passed (since they weren't available in the data I had earlier) but the ensemble corrector will eventually use both, so this is something we're making use of and not just a placeholder

6. commenting out two places where the code crashed (on our current data) :: first, the `rms_timescale()` function was giving me a type error; for some reason the `Binned_statistic` isn't getting an integer array for its bins? I'm looking into why this is popping up, but if there's already a fix for this we should get it in before the merge
the second is just that we don't have quality or error estimations to save to the FITS file (yet)

Miklnl committed:
Initial commit of cbv branch

Miklnl committed:
Merge remote-tracking branch 'origin/devel' into cbv

Miklnl committed:
Putting CBV code on pipeline form

Miklnl committed:
Merge remote-tracking branch 'origin/devel' into cbv

Miklnl committed:
Added CBV option to tess_corr and made small bug fixes to CBV code

Miklnl committed:
Included calculation of weights, update to BIC calculation

Miklnl committed:
Added option of specifying CCD and Camera

Miklnl committed:
Merge remote-tracking branch 'origin/devel' into cbv

Miklnl committed:
Auto stash before merge of "cbv" and "origin/devel"

Miklnl committed:
Added robust SNR test based on MAD

Miklnl committed:
Added axis to MAD calculation

Miklnl committed:
Added catch of existing logger handler

Miklnl committed:
Remove bad quality data before running CBV calculation

Miklnl committed:
Update weight map computation

DerekAstro committed:
Updates including comments, removal of some unused material, and minor bug fixes

DerekAstro committed:
Updates including comments, removal of some unused material, and minor bug fixes

DerekAstro committed:
Updates including comments, removal of some unused material, and minor bug fixes

DerekAstro committed:
Updates including comments, removal of some unused material, and minor bug fixes

Fill4 committed:
Fix comment indentation in ensemble

Fill4 committed:
Fix comment indentation in ensemble

Fill4 committed:
Fix comment indentation in ensemble

Fill4 committed:
Fix comment indentation in ensemble

Fill4 committed:
Pass copy of lc to do_correction instead of reference to fix errors

Fill4 committed:
Pass copy of lc to do_correction instead of reference to fix errors

Fill4 committed:
Pass copy of lc to do_correction instead of reference to fix errors

Fill4 committed:
Pass copy of lc to do_correction instead of reference to fix errors

Fill4 committed:
Add datasource argument to search when building ensemble

Fill4 committed:
Add datasource argument to search when building ensemble

Fill4 committed:
Add datasource argument to search when building ensemble

Fill4 committed:
Removed line to fix error with fits writing

Fill4 committed:
Removed line to fix error with fits writing

Fill4 committed:
Removed line to fix error with fits writing

Fill4 committed:
Added time measurements in ensemble correction

Fill4 committed:
Added time measurements in ensemble correction

Fill4 committed:
Added time measurements in ensemble correction

Fill4 committed:
Added time measurements in ensemble correction

Fill4 committed:
Added source argument to run_tesscorr and get_task

Fill4 committed:
Added source argument to run_tesscorr and get_task

Fill4 committed:
Added source argument to run_tesscorr and get_task

Fill4 committed:
Catch out of range access to ensemble_array and throw exception

Fill4 committed:
Catch out of range access to ensemble_array and throw exception

Fill4 committed:
Catch out of range access to ensemble_array and throw exception

Fill4 committed:
Catch out of range access to ensemble_array and throw exception

Fill4 committed:
Temporary fix to python rewritting memory for no reason

Fill4 committed:
Temporary fix to python rewritting memory for no reason

Fill4 committed:
Temporary fix to python rewritting memory for no reason

Fill4 committed:
Added debug plots to check ensemble stars and correction

Fill4 committed:
Added debug plots to check ensemble stars and correction

Fill4 committed:
Added debug plots to check ensemble stars and correction

Fill4 committed:
Added debug plots to check ensemble stars and correction

Miklnl committed:
Update of CBV code to now run in parallel

rhandberg committed:
Update of run_cbvprep command-line interface

Miklnl committed:
clean CBV util

Miklnl committed:
Clean-up and change in use of CBV snr value

Miklnl committed:
CBV SNR now only returns indices, function added to remove cols in CBV

rhandberg committed:
Better logging behavior when running several processes

Miklnl committed:
Added options for SNR/entropy/and number components

Miklnl committed:
Merge remote-tracking branch 'origin/cbv' into cbv

Miklnl committed:
Auto stash before merge of "cbv" and "origin/cbv"

Miklnl committed:
added missing "not" to tdqm

Miklnl committed:
Added option for making initial plots

Miklnl committed:
plotting stuff

Miklnl committed:
added "dill" (for now) to requirements

rhandberg committed:
Separated data_folder into subdirs for the different correctors

rhandberg committed:
Small cleanup of whitespace etc

rhandberg committed:
Moved some input-parameters from __init__ to subfunctions

rhandberg committed:
Changed to using dill in general instead of pickle

rhandberg committed:
Code cleanup

rhandberg committed:
Keep CBVs in memory between calls to do_correction

Miklnl committed:
Loading priors to memory in do_correction

rhandberg committed:
Added plotting utilities

rhandberg committed:
Switched more to plotting utilities

Miklnl committed:
Priors now loaded as part of CBV object

rhandberg committed:
Attempt at fixing race-condition when running multiprocessing

Miklnl committed:
Merge remote-tracking branch 'origin/cbv' into cbv

rhandberg committed:
Attempt to fix matplotlib TkAgg vs Agg backend problems

rhandberg committed:
Merge branch 'devel' into cbv

Miklnl committed:
Updating weights calculation

rhandberg committed:
Attempt at including ManualExclude filters

Maybe these should be specific for CBV?

Miklnl committed:
Testing weights stuff

Miklnl committed:
Merge remote-tracking branch 'origin/cbv' into cbv

rhandberg committed:
Reverted Mikkels local changes

Miklnl committed:
update to weighting

Miklnl committed:
Merge remote-tracking branch 'origin/cbv' into cbv

rhandberg committed:
Only save auxiliary files when running in DEBUG mode

Miklnl committed:
Incorporation of distance calculation using BallTree

Miklnl committed:
Merge remote-tracking branch 'origin/cbv' into cbv

rhandberg committed:
Added MPI scheduler for large-scale runs of tesscorr

Miklnl committed:
Integrated use of KDE priors

rhandberg committed:
Fix for uncertainty propergation

rhandberg committed:
Removed dependency on dill and updated requirements

Miklnl committed:
Update to fitting CBVs

Miklnl committed:
Merge remote-tracking branch 'origin/cbv' into cbv

Miklnl committed:
Adding S1 coarse pointing to man exclude

Miklnl committed:
Check for existence of priors and inifit + clean-up

Miklnl committed:
Update to manual exclude

Miklnl committed:
Testing spike cbvs

Miklnl committed:
change quality flag for CBV correction

Miklnl committed:
Added CBV bitmask

Miklnl committed:
Using new removal of bad points

Miklnl committed:
Updated test

Miklnl committed:
Test run on correction

Miklnl committed:
Updates to LC validation

Miklnl committed:
update

Miklnl committed:
update

Miklnl committed:
Update

Miklnl committed:
remove s1 man excludes

lmcarboneau committed:
Updates to ensemble algorithm, debugging options

lmcarboneau committed:
Updates to ensemble algorithm, debugging options

lmcarboneau committed:
Updates to ensemble algorithm, debugging options

lmcarboneau committed:
Updates to ensemble algorithm, debugging options

lmcarboneau committed:
Suppress outputs that prevent execution

lmcarboneau committed:
Suppress outputs that prevent execution

lmcarboneau committed:
Suppress outputs that prevent execution

lmcarboneau committed:
Suppress outputs that prevent execution

Miklnl committed:
Testing spike cbv separation

Miklnl committed:
Adding spike sep

Miklnl committed:
Option for datasource in run

Miklnl committed:
Implementing spike sep function

Miklnl committed:
inifit with spike cbvs

Miklnl committed:
removed old file

Miklnl committed:
No real clue of change, but file isn't used anymore anyway

Miklnl committed:
bug fix to account for spike coefficients

Miklnl committed:
Adding fitting with priors

Miklnl committed:
Better option for specifying output folder

Miklnl committed:
Update to fitting

Miklnl committed:
Added extra MAD function

Miklnl committed:
Better treatment of dimensions in distance map

Miklnl committed:
tests

Miklnl committed:
added output folder specification

lmcarboneau committed:
Updated ensemble algorithm, changes to base class

Putting it here so it's available, but future changes are coming soon; there are some (possibly) unmergeable changes in BaseCorrector that need to be addressed still. Issues/WIP incoming

lmcarboneau committed:
Updated ensemble algorithm, changes to base class

Putting it here so it's available, but future changes are coming soon; there are some (possibly) unmergeable changes in BaseCorrector that need to be addressed still. Issues/WIP incoming

lmcarboneau committed:
Updated ensemble algorithm, changes to base class

Putting it here so it's available, but future changes are coming soon; there are some (possibly) unmergeable changes in BaseCorrector that need to be addressed still. Issues/WIP incoming

lmcarboneau committed:
Updated ensemble algorithm, changes to base class

Putting it here so it's available, but future changes are coming soon; there are some (possibly) unmergeable changes in BaseCorrector that need to be addressed still. Issues/WIP incoming

Miklnl committed:
Updates to additional headers and clean-up

lmcarboneau committed:
Revert "Updated ensemble algorithm, changes to base class"

This reverts commit c774771fc081c42fc16868d85e251f517a8462e2.

lmcarboneau committed:
Revert "Updated ensemble algorithm, changes to base class"

This reverts commit c774771fc081c42fc16868d85e251f517a8462e2.

lmcarboneau committed:
Revert "Updated ensemble algorithm, changes to base class"

This reverts commit c774771fc081c42fc16868d85e251f517a8462e2.

lmcarboneau committed:
Revert "Updated ensemble algorithm, changes to base class"

This reverts commit c774771fc081c42fc16868d85e251f517a8462e2.

lmcarboneau committed:
Updated ensemble algorithm, changes to base class

lmcarboneau committed:
Updated ensemble algorithm, changes to base class

lmcarboneau committed:
Updated ensemble algorithm, changes to base class

lmcarboneau committed:
Updated ensemble algorithm, changes to base class

lmcarboneau committed:
adding user specified output_folder option

lmcarboneau committed:
adding user specified output_folder option

lmcarboneau committed:
adding user specified output_folder option

lmcarboneau commented:
TravisCI is probably going to fail because the tests need to be updated. 
Current TO-DO list looks something like:

- [ ] update tests to include the `BaseCorrector` changes, pending the code review comments
- [ ] update the ensemble algorithm one more time, which will include error estimates
- [ ] determine whether or not quality flags will be included in this release version (update if needed)
- [ ] fix whatever weirdness is going on with the rms_hour() calculation (I think something is going sideways in scipy, but I haven't found the root of the issue yet)

rhandberg committed:
Merge branch 'cbv' into devel

rhandberg committed:
Updated requirements and dropped Python 2.7 checking

rhandberg committed:
Cleanup of run_tesscorr

rhandberg committed:
Reenabled run_tessphot output folder

rhandberg committed:
Cleanup

rhandberg committed:
More cleanup

rhandberg committed:
Reintroduces datasource input

It had gone missing in all the merges

Miklnl committed:
adding datasource

Miklnl committed:
Added datasource (again), and I hate GitKraken!

Miklnl committed:
ready for running

Miklnl committed:
change of cbvarea input format

Miklnl committed:
Change to datasource input

Miklnl committed:
change again

Miklnl committed:
try again

Miklnl committed:
small changes to database search

Miklnl committed:
Adding requirement to approved datavalidation

Miklnl committed:
correcting prepare cbv for one thread

Miklnl committed:
adding datasource, camera, and ccd option to mpi tesscorr

Miklnl committed:
adding get_num_tasks function

Miklnl committed:
adding mpi4py

Miklnl committed:
remove input for summary file

Miklnl committed:
update to task count

Miklnl committed:
update num count

Miklnl committed:
update num count

Miklnl committed:
update num count

Miklnl committed:
data folder for cbv update

Miklnl committed:
update function naming

Miklnl committed:
Updating small error message

Miklnl committed:
updating file name against method and data folder

rhandberg commented:
Hi @lmcarboneau ,

Good work! Sorry it has taken me this long to respond...

I can see that most of the changes you have made to the "not ensemble" part of the code, we had actually already implemented in the CBV branch... That is now merged into the "devel" branch, so may I suggest that you discard the changes you made to the "not ensemble" parts and merge in the newest version of the devel branch... Then those things will no longer be in conflict...

After that, there are only a few small things, that needs renaming, but we can take those things after the above mentioned step...

Miklnl committed:
Change file permission of copied file

Miklnl committed:
cleanup

Miklnl committed:
adding starid as constraint in tm

lmcarboneau committed:
Merge branch 'ensemble_comments' of https://github.com/tasoc/corrections into ensemble_comments

Miklnl committed:
Include corr_status Error for target search

lmcarboneau committed:
merging and fixing conflicts

ojhall94 commented on cbv_spiketest.py:
```diff
@@ -0,0 +1,181 @@
+
+cbv_area = 334
+#data_folder= '/home/mikkelnl/ownCloud/Documents/Asteroseis/TESS/TASOC_code/corrections/corrections/data/cbv_old'
+data_folder= '/home/mikkelnl/ownCloud/Documents/Asteroseis/TESS/TASOC_code/corrections/corrections/data/cbv'
```
Is this code (`cbv_spiketest.py`) just a test on a local machine for @Miklnl ? If it is we should probably flag it as such in the docstring at the top of the script.

rhandberg replied:
Yes... But it is fixed in the newest version of the devel branch.... The issue is really that this PR should not change anything not having something to do with the ensemble code... So any changes to the CBV code, `BaseCorrector`, `TaskManager` or `run_tesscorr` should properly be reverted

ojhall94 replied:
Got it!

ojhall94 commented on corrections/BaseCorrector.py:
```diff
@@ -394,23 +424,41 @@ def save_lightcurve(self, lc, output_folder=None):
 				hdu['LIGHTCURVE'].data['FLUX_CORR'] = lc.flux
-				hdu['LIGHTCURVE'].data['FLUX_CORR_ERR'] = lc.flux_err
-				hdu['LIGHTCURVE'].data['QUALITY'] = lc.quality
+				#hdu['LIGHTCURVE'].data['FLUX_CORR_ERR'] = lc.flux_err
```
Are these supposed to remain commented out in the final version?

rhandberg replied:
No... Again, an example of a change in this PR that we should revert

ojhall94 commented on corrections/BaseCorrector.py:
```diff
@@ -422,7 +470,7 @@ def save_lightcurve(self, lc, output_folder=None):
 
 				# Save the updated FITS file:
-				hdu.flush()
+				# hdu.flush()
```
Is this supposed to remain commented out in the final version?

ojhall94 commented on corrections/BaseCorrector.py:
```diff
@@ -457,4 +505,4 @@ def save_lightcurve(self, lc, output_folder=None):
 		save_file = os.path.relpath(save_file, output_folder).replace('\\', '/')
 
-		return save_file
+		return save_file
```
Is there a little weird symbol for everyone else too?

ojhall94 commented on corrections/ensemble.py:
```diff
@@ -47,29 +70,43 @@ def do_correction(self, lc):
-        # NOTE: Is frange supposed to be missing the brackets and not join the percentiles ??
-        frange = np.percentile(lc.flux, 95) - np.percentile(lc.flux, 5) / np.mean(lc.flux)
+        lc_quality_mask = (lc.quality == 0)
+        # lc.time = lc.time[lc_quality_mask]
```
Should these remain commented out in the final version?

ojhall94 commented on corrections/ensemble.py:
```diff
@@ -97,208 +135,237 @@ def do_correction(self, lc):
-                # Compute the rest of its data. NOTE: Change this to the database or not ???
-                frange = np.percentile(next_star_lc.flux, 95) - np.percentile(next_star_lc.flux, 5) / np.mean(next_star_lc.flux)
+                next_star_lc_quality_mask = (next_star_lc.quality == 0)
+                # next_star_lc.time = next_star_lc.time[next_star_lc_quality_mask]
```
Should these remain commented out in the final version?

ojhall94 commented on .travis.yml:
```diff
@@ -1,6 +1,5 @@
 language: python
 python:
```
`.travis.yml` should be checked out from the PR.

ojhall94 commented on cbv_spiketest.py:
```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env python3
```
`cbv_spiketest.py` should be checked out from the PR

ojhall94 commented on corrections/__init__.py:
```diff
@@ -4,6 +4,7 @@
 from __future__ import absolute_import
```
`__init__.py` should be checked out from the PR

ojhall94 commented on corrections/cbv_corrector/CBVCorrector.py:
```diff
@@ -0,0 +1,775 @@
+#!/usr/bin/env python
```
`CBVCorrector.py`  should be checked out from the PR

ojhall94 commented on corrections/cbv_corrector/GOC_code.py:
```diff
@@ -0,0 +1,456 @@
+#!/usr/bin/env python3
```
`GOC_code.py`  should be checked out from the PR

ojhall94 commented:
Right I have more of a grasp on the situation now. Something seems to have gone wrong during the rebase: all files apart from the EnsembleCorrector content should be up to date with the latest in `devel`, but all files are still being flagged as having changes with respect to `devel`.

I think the right course of action would be a merge `devel` -> `ensemble_comments` and in cases of merge conflicts prefer the `devel` version in all cases except for the Ensemble code content.

I'll have a swing at this tomorrow morning.

Commit:
Reset the rebase to most recent devel status

I have unstaged BaseCorretor and run_tesscor sot hat only ensemble.py is changed from the most current status of devel

Commit:
Merge remote-tracking branch 'origin/ensemble_comments' into ensemble_comments

ojhall94 commented:
Okay, something funky is happening. The only differences between the current state of `devel` and `ensemble_comments` are:

`corrections/ensemble.py`
`corrections/BaseCorrector.py`
`run_tesscor.py`

Yet it insists on flagging 24 files as changed with respect to devel.
I've done my best to remedy whats going on but its gone beyond my Git knowledge. I think its time to abandon branch (@lmcarboneau) for a new PR off the devel branch and just refer back to this thread as the commit history on the ensemble work.

rhandberg commented:
I think I second that @ojhall94... It seems that would be the easiest way out, as this also goes way beyond my Git-knowledge! So a new PR with just the changes to `ensemble.py`, based off the current devel branch.

rhandberg commented:
This PR has been bypassed by https://github.com/tasoc/corrections/pull/14 and is therefore closed.
